### ISSUE TITLE
key is not type-checked on `template`

### ISSUE THREAD

**Issue by brc-dd** [vuejs/language-tools#4539]
### Vue - Official extension or vue-tsc version

2.0.26

### VSCode version

1.92.0-insider

### Vue version

3.4.31

### TypeScript version

5.5.3

### Steps to reproduce

1. `git clone git@github.com:brc-dd/volar-template-key-bug.git`
2. `pnpm i`
3. `pnpm vue-tsc --noEmit`

### What is expected?

Error on both line 4 and line 5 of `Foo.vue`

### What is actually happening?

Error is only on line 5. There should be an error saying `Type '{}' is not assignable to type 'PropertyKey | undefined'.` for `:key` prop of `<template>` too.

### Link to minimal reproduction

https://github.com/brc-dd/volar-template-key-bug

**Comment by davidmatter** [comment]
**Input**
```vue
<script setup lang="ts"></script>

<template>
  <template v-if="true" :key="">
	<div></div>
  </template>
  <template v-if="true" :key="{}">
	<div></div>
  </template>

  <div v-if="true" :key=""></div>
  <div v-if="true" :key="{}"></div>
</template>
```

**Output**
```ts
// ...
if (true) {
__VLS_elementAsFunction(__VLS_intrinsicElements.div, __VLS_intrinsicElements.div)({});
}
if (true) {
__VLS_elementAsFunction(__VLS_intrinsicElements.div, __VLS_intrinsicElements.div)({});
}
if (true) {
__VLS_elementAsFunction(__VLS_intrinsicElements.div, __VLS_intrinsicElements.div)({key: ((__VLS_ctx.)), });
// @ts-ignore
[,];
}
if (true) {
__VLS_elementAsFunction(__VLS_intrinsicElements.div, __VLS_intrinsicElements.div)({key: (({})), });
}
// ...
```

**Comment by so1ve** [comment]
https://github.com/vuejs/core/blob/93324b2ec06b26662b77abc2a75d21ecbe8913db/packages/compiler-core/src/transforms/vIf.ts#L206-L216

A `<template>` with v-if will be turned into an `IF_BRANCH` node, we should handle it in https://github.com/vuejs/language-tools/blob/master/packages/language-core/lib/codegen/template/elementDirectives.ts.

**Comment by brc-dd** [comment]
Ah, it's not working for `v-for` in template's either. I had put v-if for easier repro.

**Comment by so1ve** [comment]
Should be fixed by https://github.com/vuejs/language-tools/commit/cb540323d4f0e80feed4feb4fea301f7d532f0c5? Do you mean there is no type checking?

**Comment by brc-dd** [comment]
> Do you mean there is no type checking?

Yeah it's not there for key. This should've shown errors: 

```vue
<script setup lang="ts"></script>

<template>
  <template v-for="x in [1]" :key="{}">
    <div>{{ x }}</div>
  </template>
</template>
```

**Comment by so1ve** [comment]
Got it :open_hands:

**Comment by so1ve** [comment]
My PR doesn't report `v-bind is missing expression`, since `CompilerDOM.compile()` itself doesn't throw on those invalid directives.

**Comment by brc-dd** [comment]
Ah yeah. It can be left I guess. I'll track it on the core repo. Looks like compiler-dom is generating invalid JS for  `<template v-for="x in [1]" :key="" />` without complaining.

**Comment by so1ve** [comment]
Yeah, `v-if` with `:key` generates `{ key: 0 }` while `v-for` with `:key` generates `{ key: }`
